[CLI] Add 'mlflow runs link-traces' command

nsthorat opened the pull request:
### Related Issues/PRs

N/A

### What changes are proposed in this pull request?

This PR adds a new CLI command `mlflow runs link-traces` to link traces to runs, enabling users to establish relationships between traces and runs within the same experiment via the command line interface.

- Added `link-traces` command to `mlflow/runs.py` that allows linking one or more traces to an existing run
- Supports both long form (`--trace-id`) and short form (`-t`) options for specifying trace IDs  
- Can link up to 100 traces in a single command (API limitation documented in help text)
- Returns a success message with the count of traces linked
- Added comprehensive test coverage in `tests/test_runs.py`

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

#### New Unit Tests
Added extensive test coverage in `tests/test_runs.py`:
- `test_link_traces_single_trace`: Tests linking a single trace
- `test_link_traces_multiple_traces`: Tests linking multiple traces
- `test_link_traces_with_short_option`: Tests using the `-t` short option
- `test_link_traces_file_store_error`: Tests error handling for unsupported FileStore
- `test_link_traces_too_many_traces_error`: Tests error when exceeding 100 trace limit
- `test_link_traces_missing_run_id`: Tests validation for missing run ID
- `test_link_traces_missing_trace_id`: Tests validation for missing trace IDs
- `test_link_traces_generic_error`: Tests generic error handling

#### Manual Testing
Tested manually with a Databricks tracking server:

1. Created a new run:
```bash
export DATABRICKS_HOST=https://[REDACTED].cloud.databricks.com
export DATABRICKS_TOKEN=[REDACTED]
export MLFLOW_TRACKING_URI=databricks
export MLFLOW_EXPERIMENT_ID=2178582188830602

mlflow runs create --experiment-id $MLFLOW_EXPERIMENT_ID --run-name "test-link-traces-cli"
# Output: {"run_id": "37ba9f983ca840a08d40f2d5750c7939", ...}
```

2. Found existing traces in the experiment:
```bash
mlflow traces search --experiment-id 2178582188830602 --max-results 5
# Listed traces including: tr-6fb236e9e5ecdc6b2de7e3fbe6857137
```

3. Linked a single trace to the run:
```bash
mlflow runs link-traces --run-id 37ba9f983ca840a08d40f2d5750c7939 --trace-id tr-6fb236e9e5ecdc6b2de7e3fbe6857137
# Output: Successfully linked 1 trace(s) to run '37ba9f983ca840a08d40f2d5750c7939'
```

4. Linked multiple traces using short option:
```bash
mlflow runs link-traces --run-id 37ba9f983ca840a08d40f2d5750c7939 -t tr-10cb247de042863645f08fcb56d868b3 -t tr-8086573c30ff0cc4163bdb5d52781436
# Output: Successfully linked 2 trace(s) to run '37ba9f983ca840a08d40f2d5750c7939'
```

5. Verified traces are linked by searching with run_id filter:
```bash
mlflow traces search --experiment-id 2178582188830602 --filter-string "run_id = '37ba9f983ca840a08d40f2d5750c7939'" --max-results 10
# Showed all 3 linked traces
```

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [x] API references
  - [ ] Instructions

Updated docstrings and CLI help text in the code. Full documentation updates may be needed as a follow-up.

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Added a new CLI command `mlflow runs link-traces` to link traces to runs, allowing users to establish relationships between traces and runs within the same experiment.

Example usage:
```bash
# Link a single trace
mlflow runs link-traces --run-id <run-id> --trace-id <trace-id>

# Link multiple traces
mlflow runs link-traces --run-id <run-id> --trace-id <trace1> --trace-id <trace2>

# Using short option
mlflow runs link-traces --run-id <run-id> -t <trace1> -t <trace2>

# Using short options
mlflow runs link-traces -x <exp-id> --run-id <run-id> -t <trace1> -t <trace2>
```

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [x] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

nsthorat committed:
Add 'mlflow runs link-traces' CLI command

This commit adds a new CLI command to link traces to runs, enabling users
to establish relationships between traces and runs within the same experiment
via the command line.

- Added 'link-traces' command to mlflow/runs.py with support for linking
  single or multiple traces to a run (maximum 100 traces)
- Added extensive test coverage in tests/test_runs.py
- Supports both long form (--trace-id) and short form (-t) options

The command uses MlflowClient.link_traces_to_run() which requires a
database-backed tracking store (not supported in FileStore).

Signed-off-by: Nikhil Thorat <nikhil.thorat@databricks.com>
Signed-off-by: nsthorat <nsthorat@gmail.com>

nsthorat committed:
Add experiment-id flag to link-traces command

- Added optional --experiment-id / -x flag to set experiment context
- Updated tests to use inline return_value pattern for cleaner code
- Added test coverage for experiment-id functionality

Signed-off-by: nsthorat <nsthorat@gmail.com>

nsthorat committed:
Fix ruff formatting and remove unused variables

- Removed unused mock_client_class variables in tests
- Applied ruff formatting to test file

Signed-off-by: nsthorat <nsthorat@gmail.com>

nsthorat committed:
Remove unnecessary experiment-id flag

The experiment-id flag is not needed since the backend trace-to-run
association is a simple 1:1 mapping without experiment context.
Setting experiment context doesn't affect the linking operation.

- Removed --experiment-id / -x option from link-traces command
- Updated function signature to remove experiment_id parameter
- Simplified implementation by removing mlflow.set_experiment call
- Removed test_link_traces_with_experiment_id test
- Updated remaining tests to remove experiment-id related assertions

Signed-off-by: nsthorat <nsthorat@gmail.com>

nsthorat commented on mlflow/runs.py:
```diff
@@ -206,3 +207,53 @@ def create_run(
+
+    try:
+        # Set the experiment context if provided
+        if experiment_id:
```
Ah I didn't realize that, excellent, removed (actually claude knew this, and I didn't and I was like where's your experiment id dude?!)

BenWilson2 replied:
If I had a dime for every time this has happened to me in the last month.... ;)

BenWilson2 reviewed (approved):
looks good!

harupy commented on tests/test_runs.py:
```diff
@@ -212,3 +214,113 @@ def test_create_run_duplicate_tag_key():
+        "Linking traces to runs is not supported in FileStore. "
+        "Please use a database-backed store (e.g., SQLAlchemy store) for this feature."
+    )
+    with patch("mlflow.runs.MlflowClient", return_value=mock_client):
```
```suggestion
    with patch("mlflow.runs.MlflowClient.link_traces_to_run", side_effect = MlflowException(
        "Linking traces to runs is not supported in FileStore. "
        "Please use a database-backed store (e.g., SQLAlchemy store) for this feature."
    )):
```

harupy replied:
can you apply this change to other `mock.patch`?

nsthorat replied:
Done

harupy commented on mlflow/runs.py:
```diff
@@ -206,3 +207,41 @@ def create_run(
+    linked to runs to establish relationships between traces and runs.
+    Maximum 100 traces can be linked in a single command.
+    """
+    # Convert tuple to list
+    trace_ids = list(trace_id)
+
+    try:
+        client = MlflowClient()
+        client.link_traces_to_run(trace_ids, run_id)
```
```suggestion
    try:
        client = MlflowClient()
        client.link_traces_to_run(list(trace_ids), run_id)
```

nsthorat replied:
Done!

nsthorat committed:
Address harupy's review comments

- Specify parameter name for --trace-id click option
- Change function parameter from trace_id to trace_ids
- Inline list(trace_ids) conversion in function body
- Use direct method patching instead of class patching in tests
- Remove unused MagicMock import from tests

Signed-off-by: nsthorat <nsthorat@gmail.com>

nsthorat commented on mlflow/runs.py:
```diff
@@ -206,3 +207,41 @@ def create_run(
+    help="ID of the run to link traces to.",
+)
+@click.option(
+    "--trace-id",
```
Done

nsthorat replied:
done

nsthorat commented on mlflow/runs.py:
```diff
@@ -206,3 +207,41 @@ def create_run(
+    required=True,
+    help="Trace ID to link to the run. Can be specified multiple times (maximum 100 traces).",
+)
+def link_traces(run_id: str, trace_id: tuple[str, ...]) -> None:
```
done

harupy reviewed (approved):
LGTM!
